File: src/main/java/org/cloudfoundry/credhub/controller/v1/CertificatesController.java
package org.cloudfoundry.credhub.controller.v1;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import org.apache.commons.lang3.StringUtils;
import org.cloudfoundry.credhub.audit.CEFAuditRecord;
import org.cloudfoundry.credhub.audit.OperationDeviceAction;
import org.cloudfoundry.credhub.audit.entity.GetCertificateByName;
import org.cloudfoundry.credhub.audit.entity.RegenerateCertificate;
import org.cloudfoundry.credhub.audit.entity.UpdateTransitionalVersion;
import org.cloudfoundry.credhub.handler.CertificatesHandler;
import org.cloudfoundry.credhub.request.CertificateRegenerateRequest;
import org.cloudfoundry.credhub.request.UpdateTransitionalVersionRequest;
import org.cloudfoundry.credhub.view.CertificateCredentialsView;
import org.cloudfoundry.credhub.view.CertificateView;
import org.cloudfoundry.credhub.view.CredentialView;

import static org.cloudfoundry.credhub.controller.v1.CertificatesController.API_V1_CERTIFICATES;

@RestController
@RequestMapping(API_V1_CERTIFICATES)
public class CertificatesController {

  public static final String API_V1_CERTIFICATES = "api/v1/certificates";

  private final CEFAuditRecord auditRecord;
  private final CertificatesHandler certificatesHandler;

  @Autowired
  public CertificatesController(final CertificatesHandler certificateHandler, final CEFAuditRecord auditRecord) {
    super();
    this.certificatesHandler = certificateHandler;
    this.auditRecord = auditRecord;
  }

  @PostMapping(value = "/{certificateId}/regenerate", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
  @ResponseStatus(HttpStatus.OK)
  public CredentialView regenerate(
    @RequestBody(required = false) final CertificateRegenerateRequest requestBody,
    @PathVariable final String certificateId
  ) {
    final CertificateRegenerateRequest finalRequestBody = requestBody == null ? new CertificateRegenerateRequest() : requestBody;

    final RegenerateCertificate certificate = new RegenerateCertificate();
    certificate.setTransitional(finalRequestBody.isTransitional());
    auditRecord.setRequestDetails(certificate);

    return certificatesHandler.handleRegenerate(certificateId, finalRequestBody);
  }

  @GetMapping(value = "", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
  @ResponseStatus(HttpStatus.OK)
  public CertificateCredentialsView getAllCertificates() {
    auditRecord.setRequestDetails(() -> OperationDeviceAction.GET_ALL_CERTIFICATES);

    return certificatesHandler.handleGetAllRequest();
  }

  @PutMapping(value = "/{certificateId}/update_transitional_version", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
  @ResponseStatus(HttpStatus.OK)
  public List<CertificateView> updateTransitionalVersion(@RequestBody final UpdateTransitionalVersionRequest requestBody,
                                                         @PathVariable final String certificateId) {
    final UpdateTransitionalVersion details = new UpdateTransitionalVersion();
    details.setVersion(requestBody.getVersionUuid());
    auditRecord.setRequestDetails(details);
    return certificatesHandler.handleUpdateTransitionalVersion(certificateId, requestBody);
  }

  @GetMapping(value = "", params = "name", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
  @ResponseStatus(HttpStatus.OK)
  public CertificateCredentialsView getCertificateByName(@RequestParam("name") final String name) {
    final String credentialNameWithPrependedSlash = StringUtils.prependIfMissing(name, "/");
    final GetCertificateByName details = new GetCertificateByName();
    details.setName(name);
    auditRecord.setRequestDetails(details);
    return certificatesHandler.handleGetByNameRequest(credentialNameWithPrependedSlash);
  }
}

File: backends/credhub/src/main/kotlin/org/cloudfoundry/credhub/certificates/CertificatesController.kt
package org.cloudfoundry.credhub.certificates

import org.apache.commons.lang3.StringUtils
import org.cloudfoundry.credhub.audit.CEFAuditRecord
import org.cloudfoundry.credhub.audit.OperationDeviceAction
import org.cloudfoundry.credhub.audit.entities.GetCertificateByName
import org.cloudfoundry.credhub.audit.entities.RegenerateCertificate
import org.cloudfoundry.credhub.audit.entities.UpdateTransitionalVersion
import org.cloudfoundry.credhub.requests.CertificateRegenerateRequest
import org.cloudfoundry.credhub.requests.CreateVersionRequest
import org.cloudfoundry.credhub.requests.UpdateTransitionalVersionRequest
import org.cloudfoundry.credhub.views.CertificateCredentialsView
import org.cloudfoundry.credhub.views.CertificateView
import org.cloudfoundry.credhub.views.CredentialView
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping(path = [CertificatesController.ENDPOINT], produces = [MediaType.APPLICATION_JSON_UTF8_VALUE])
class CertificatesController(
    val certificatesHandler: CertificatesHandler,
    val auditRecord: CEFAuditRecord
) {

    companion object {
        const val ENDPOINT = "/api/v1/certificates"
    }

    @RequestMapping(method = [RequestMethod.GET], path = [""])
    @ResponseStatus(HttpStatus.OK)
    fun getAllCertificates(): CertificateCredentialsView {
        auditRecord.setRequestDetails { OperationDeviceAction.GET_ALL_CERTIFICATES }

        return certificatesHandler.handleGetAllRequest()
    }

    @RequestMapping(method = [RequestMethod.GET], path = [""], params = ["name"])
    @ResponseStatus(HttpStatus.OK)
    fun getCertificateByName(@RequestParam("name") name: String): CertificateCredentialsView {
        val credentialName = StringUtils.prependIfMissing(name, "/")
        val details = GetCertificateByName()
        details.name = name
        auditRecord.requestDetails = details

        return certificatesHandler.handleGetByNameRequest(credentialName)
    }

    @RequestMapping(method = [RequestMethod.POST], path = ["/{certificateId}/regenerate"])
    @ResponseStatus(HttpStatus.OK)
    fun regenerateCertificate(
        @PathVariable("certificateId") certificateId: String,
        @RequestBody(required = false) requestBody: CertificateRegenerateRequest?
    ): CredentialView {
        val finalRequestBody = requestBody ?: CertificateRegenerateRequest()
        val certificate = RegenerateCertificate()
        certificate.transitional = finalRequestBody.isTransitional
        auditRecord.requestDetails = certificate

        return certificatesHandler.handleRegenerate(certificateId, finalRequestBody)
    }

    @RequestMapping(method = [RequestMethod.PUT], path = ["/{certificateId}/update_transitional_version"])
    @ResponseStatus(HttpStatus.OK)
    fun updateTransitionalVersion(
        @PathVariable("certificateId") certificateId: String,
        @RequestBody requestBody: UpdateTransitionalVersionRequest
    ): List<CertificateView> {
        val details = UpdateTransitionalVersion()
        details.version = requestBody.versionUuid
        auditRecord.requestDetails = details

        return certificatesHandler.handleUpdateTransitionalVersion(certificateId, requestBody)
    }

    @RequestMapping(method = [RequestMethod.GET], path = ["{certificateId}/versions"])
    @ResponseStatus(HttpStatus.OK)
    fun getAllCertificateVersions(
        @PathVariable("certificateId") certificateId: String,
        @RequestParam(value = "current", required = false, defaultValue = "false") current: Boolean
    ): List<CertificateView> {
        return certificatesHandler.handleGetAllVersionsRequest(certificateId, current)
    }

    @RequestMapping(method = [RequestMethod.POST], path = ["{certificateId}/versions"])
    @ResponseStatus(HttpStatus.OK)
    fun createVersion(
        @RequestBody @Validated requestBody: CreateVersionRequest,
        @PathVariable("certificateId") certificateId: String
    ): CertificateView {
        return certificatesHandler.handleCreateVersionsRequest(certificateId, requestBody)
    }

    @RequestMapping(method = [RequestMethod.DELETE], path = ["{certificateId}/versions/{versionId}"])
    @ResponseStatus(HttpStatus.OK)
    fun deleteVersion(
        @PathVariable("certificateId") certificateId: String,
        @PathVariable("versionId") versionId: String
    ): CertificateView {
        return certificatesHandler.handleDeleteVersionRequest(certificateId, versionId)
    }
}
